packages/workspace/src/browser/workspace-commands.ts

```diff
@@ -104,9 +107,11 @@ export class WorkspaceCommandContribution implements CommandContribution {
                     initialValue: uri.path.base,
                     validate: name => this.validateFileName(name, parent)
                 });
-                dialog.open().then(name =>
-                    this.fileSystem.move(uri.toString(), uri.parent.resolve(name).toString())
-                );
+                dialog.open().then(name => {
+                    this.fileSystem.move(uri.toString(), uri.parent.resolve(name).toString()).then(() => {
```

Comment:
I think @svenefftinge meant that you only should close an existing editor and open new:
```ts
const newUri = uri.parent.resolve(name);
this.fileSystem.move(uri.toString(), newUri.toString()).then(() => {
  const widget = this.editorService.getEditor(uri);
  if (editor) {
    const selection = widget.editor.selection;
    widget.close();
    open(this.openerService, newUri, {selection});
  }
});
```

Ideally, it should be done not here but on move/rename events from the filesystem. Making it here:
- breaks the unidirectional flow, only the filesystem should be the source of truth;
- cover only a specific case, e.g. renaming files in Finder won't make the same effect;
- make the workspace extension depend on the editor extension.

Reply:
No, I did not mean that a real rename would be better. But we have to take care of all the loose ends.

Reply:
Maybe we should have an generic key to reference the widget directly to avoid lose ends ?

Reply:
Could you elaborate on what you mean?

Reply:
I mean that if something has to keep a reference to an editor widget it should reference the widget object directly or we should provide a key that is not bound to any resource to refer to it. Like some abstract id.

So the URI would be used as key to open a new widget or find a widget for that URI but once you get a reference to it, it should be the id or the object directly, a bit like files names and files descriptors.. you open(filename) get a file descriptor so you could getWidgetForURI() get a widget for a URI but use getWidgetId() if you already had a widget and now know only it's id. I don't see where however that would be useful.. just keeping the object seems ok since serializing that doesn't make sense as the widget-manager is in the frontend.

>Modifying a widget after creation will break the contract of the widget manager. After reload it will try to create a widget based on the old uri.

Why would it try to reload with an old uri ? if you updated it before you serialized the widget ?

Granted if it's changed between save and restore there's nothing we can do but that is true whatever we do...